src/nlp/search.py

```diff
@@ -0,0 +1,366 @@
+import logging
+from typing import TYPE_CHECKING, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from tqdm.auto import tqdm
+
+
+if TYPE_CHECKING:
+    from .arrow_dataset import Dataset  # noqa: F401
+
+
+try:
+    import elasticsearch as es
+    import elasticsearch.helpers
+
+    _has_elasticsearch = True
+except ImportError:
+    _has_elasticsearch = False
+
+try:
+    import faiss
+
+    _has_faiss = True
+except ImportError:
+    _has_faiss = False
+
+
+logger = logging.getLogger(__name__)
+
+
+class MissingIndex(Exception):
+    pass
+
+
+class BaseIndex:
+    """Base class for indexing"""
+
+    def search(self, query, k: int = 10) -> Tuple[List[float], List[int]]:
```

Comment:
In this particular case the signature of `search` is a convention widely used so I'm not sure it will change.
However I agree that using NamedTuples can make it more readable as a documentation.

Reply:
Agree, NamedTuples have 0 disadvantages IMO